### PR TITLE
Add the Agglomerative Hierarchical Clustering (AHC) on manifolds.

### PR DESCRIPTION
**Pull request by YannCabanes**
Add the Agglomerative Hierarchical Clustering (AHC) on manifolds and its test.

### CODE REVIEW

File: geomstats/learning/ahc.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
+"""The Agglomerative Hierarchical Clustering (AHC) on manifolds."""
+
+from sklearn.cluster import AgglomerativeClustering
+
+import geomstats.backend as gs
+
+
+class AgglomerativeHierarchicalClustering(AgglomerativeClustering):
```
Comment (ninamiolane) [inline geomstats/learning/ahc.py:8]: Did you modify this class, w.r.t. to scikit-learn? Or are we "only" using it?

File: tests/test_ahc.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+class TestAgglomerativeHierarchicalClustering(geomstats.tests.TestCase):
+    """Class defining the Poincare polydisk tests."""
+
+    def setUp(self):
+        """Define the parameters to test."""
+        gs.random.seed(1234)
+        self.n_clusters = 2
+        self.dimension = 2
+        self.space = Euclidean(dimension=self.dimension)
```
Comment (ninamiolane) [inline tests/test_ahc.py:17]: Let's test this with a non-Euclidean metric, to see if it works with geomstats' distances? It'll be good to have it in the test, not only in the examples.

Comment (YannCabanes) [reply tests/test_ahc.py:17]: I will show that it is working well with other metrics in my next PR with an example on the sphere.
Is this fine for you ?

File: geomstats/learning/ahc.py (hunk 1 above)
Comment (YannCabanes) [reply geomstats/learning/ahc.py:8]: I made it easier to use thanks to the distance parameter instead of the affinity parameter which in scikit-learn, when callable, should be a function that takes the data as input and return a distance matrix.
In my code, we just need to specify the distance and this function is automatically defined.
Plus I changed the default parameter of the linkage parameter for it to be compatible with non-euclidean metrics.

File: tests/test_ahc.py (hunk 2 above)
Comment (YannCabanes) [reply tests/test_ahc.py:17]: Ok, I will add other metrics in the test too.

File: tests/test_ahc.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+"""Unit tests for the KNN classifier."""
```
Comment (nguigs) [inline tests/test_ahc.py:1]: Wrong docstring!

File: tests/test_ahc.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+import geomstats.backend as gs
+import geomstats.tests
+from geomstats.geometry.euclidean import Euclidean
+from geomstats.geometry.hypersphere import Hypersphere
+from geomstats.learning.ahc import AgglomerativeHierarchicalClustering
+
+
+class TestAgglomerativeHierarchicalClustering(geomstats.tests.TestCase):
+    """Class defining the Poincare polydisk tests."""
```
Comment (nguigs) [inline tests/test_ahc.py:11]: Wrong docstring

File: tests/test_ahc.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        """Define the parameters to test."""
+        gs.random.seed(1234)
+
+    @geomstats.tests.np_only
+    def test_fit_euclidean_distance(self):
+        """Test the 'fit' class method using the Euclidean distance."""
+        n_clusters = 2
+        dimension = 2
+        space = Euclidean(dimension=dimension)
+        distance = space.metric.dist
```
Comment (nguigs) [inline tests/test_ahc.py:23]: Nice! Could you duplicate this test with a string distance argument so that the condition `if isinstance(distance, str)` is met and actually tested? We would reach 100% covering then!

File: geomstats/learning/ahc.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    Recursively merges the pair of clusters that minimally increases
+    a given linkage distance.
+
+    Parameters
+    ----------
+    n_clusters : int or None, default=2
+        The number of clusters to find. It must be ``None`` if
+        ``distance_threshold`` is not ``None``.
+    distance : str or callable, default='euclidean'
```
Comment (nguigs) [inline geomstats/learning/ahc.py:19]: Cool! Could you add an exemple section in the docstring so we know how to pass a geomstats geodesic distance? As in https://github.com/geomstats/geomstats/blob/6cf2aba8f917718028e445207eb3705fb56c98ab/geomstats/learning/online_kmeans.py#L142 for example:
```
    Example
    -------
    >>> from geomstats.geometry.hypersphere import Hypersphere
    >>> from geomstats.learning.ahc import AgglomerativeHierarchicalClustering
    >>> sphere = Hypersphere(dimension=2)
    >>> X = sphere.random_uniform(10)
    >>> distance = sphere.metric.dist
    >>> AgglomerativeHierarchicalClustering(distance=distance)
    >>> clustering.fit(X)
```
